### PR TITLE
[PTen] Add compatible reshape method for Tensor

### PR DESCRIPTION
**Pull request by chenwhql**
### PR types

Function optimization
### PR changes

APIs
### Describe

[PTen] Add compatible reshape method for Tensor

https://github.com/PaddlePaddle/Paddle/pull/37122 的TODO事项之一，恢复Tensor的reshape接口支持，确保兼容，目前reshape本质上仍是底层Tensor的Resize功能，由于这里reshape接口的含义和Python端reshape接口含义有出入，所以这里的reshape将会deprecated

### CODE REVIEW

**Review by Shixiaowei02 — DISMISSED** [review]

File: paddle/pten/api/lib/tensor.cc [hunk 1]
```diff
@@ -105,9 +105,21 @@ std::vector<int64_t> Tensor::shape() const {
...
-      "The reshape operation is not supported now, "
-      "and it will be implemented by calling the reshape kernel later."));
+  LOG(WARNING) << "The function of resetting the shape of the uninitialized "
+                  "Tensor of the `reshape` method is deprecated since version "
+                  "2.3, and will be removed in version 2.4, please use "
+                  "`paddle::experimental::full` method to create a new Tensor "
+                  "instead. "
+                  "reason: `reshape` means changing the tensor shape without "
+                  "touching underlying data, this requires the total size of "
+                  "the tensor to remains constant.";
```
Comment (MingMingShangTian) [inline paddle/pten/api/lib/tensor.cc]: to remains  => to remain

File: paddle/pten/api/lib/tensor.cc [hunk 2]
```diff
@@ -105,9 +105,21 @@ std::vector<int64_t> Tensor::shape() const {
...
+  if (detail::IsDenseTensor(impl_)) {
+    std::dynamic_pointer_cast<pten::DenseTensor>(impl_)->Resize(
+        framework::make_ddim(shape));
+  } else {
+    PADDLE_THROW(platform::errors::Unimplemented(
+        "Only supported reshape operation on DenseTensor now."));
```
Comment (zyfncg) [inline paddle/pten/api/lib/tensor.cc]: support？

**Review by YuanRisheng — DISMISSED** [review]

**Review by zyfncg — DISMISSED** [review]

File: paddle/pten/api/lib/tensor.cc (hunk 1 above)
Comment (chenwhql) [reply paddle/pten/api/lib/tensor.cc]: done, thx

File: paddle/pten/api/lib/tensor.cc (hunk 2 above)
Comment (chenwhql) [reply paddle/pten/api/lib/tensor.cc]: done, thx

**Review by Shixiaowei02 — APPROVED** [review]

**Review by zyfncg — APPROVED** [review]

**Review by YuanRisheng — APPROVED** [review]

**Pull request merged**